dbms/src/Storages/Distributed/DirectoryMonitor.cpp

```diff
@@ -293,20 +310,53 @@ void StorageDistributedDirectoryMonitor::readQueryAndSettings(
                     << "It may lack support for new features.");
         }
 
+        /// Verify checksums
+        if (initiator_revision >= DBMS_MIN_REVISION_WITH_EXTRAS_CHECKSUM_IN_DIST_BATCH)
+        {
+            CityHash_v1_0_2::uint128 expected_checksum;
+            CityHash_v1_0_2::uint128 calculated_checksum;
+            size_t checksum_size = sizeof(expected_checksum)*2;
+
+            if (extra_info.available() < checksum_size)
+                throw Exception("Not enough data", ErrorCodes::CORRUPTED_DATA);
+
+            ReadBuffer checksums(extra_info.position(), extra_info.available(), extra_info.available() - checksum_size);
+            extra_info = ReadBufferFromMemory(extra_info.position(), extra_info.available() - checksum_size);
+
+            /// query checksum
+            readPODBinary(expected_checksum, checksums);
+            calculated_checksum = CityHash_v1_0_2::CityHash128(insert_query.data(), insert_query.size());
+            assertChecksum(expected_checksum, calculated_checksum);
+
+            /// extra_info checksum (all data except itself -- this checksum)
+            readPODBinary(expected_checksum, checksums);
+            calculated_checksum = CityHash_v1_0_2::CityHash128(extra_info_as_string.data(), extra_info_as_string.size() - sizeof(expected_checksum));
+            assertChecksum(expected_checksum, calculated_checksum);
+
+            if (!checksums.eof())
+                throw Exception("Unhandled checksums", ErrorCodes::LOGICAL_ERROR);
+        }
+
         insert_settings.deserialize(extra_info);
 
         /// Add handling new data here, for example:
         /// if (initiator_revision >= DBMS_MIN_REVISION_WITH_MY_NEW_DATA)
         ///    readVarUInt(my_new_data, extra_info);
+
+        if (!extra_info.eof())
+            throw Exception("Unhandled extra bytes", ErrorCodes::CORRUPTED_DATA);
```

Comment:
No, we shouldn't throw an exception here because it's valid case.
The intitiator of a distributed query can have older version so it can send more data than a shard knows how to read.
We show a warning in this case and that's enough.

Reply:
Nice catch!